### DISCUSSION TITLE
RFC: FVM Changes

### DISCUSSION THREAD

**Discussion by leoafarias** [leoafarias/fvm#469] [General]
RFC: Locked:
Thank you all for the feedback, we have made considerable improvements for 3.0.0 based on the information provided. 

A few takeaways.

- Global command stays; however, if you use global, you should be using FVM executable, not pub global.
- We have simplified the FVM configuration.
- Much better DX, notifications, warnings, and corrective actions.
- Tighter integration with VsCode.

-- Original --

## TLDR:
- FVM has come to a critical tool in the Flutter Dev toolbox.
- We should aim to improve compatibility with Flutter tools and recommended workflows where possible.
- Suggestion - Deprecate `global` command.
- Suggestion - Executables > pub global activate
- Refactor, with the simplified workflows in mind.

---

When I first created FVM, I never imagined that it would have grown in scope and usage to become the go tool for the Flutter version management challenges.

The community has embraced it in a way I did not expect, and it is great to see the number of people mentioning how much easier FVM makes their dev experience and how much it has saved them.

However, even though FVM does an excellent job in most of the cases and it's clear that it's solving a significant problem, it can, in a way, create "noise," where issues are not evident where they could be coming from (FVM, Flutter or Dev Environment).

That is an ongoing challenge when seeing new issues on both FVM and issues that mention the use of FVM within the Flutter repo.

Because FVM has become a more critical part of the Flutter developer toolset, I would like to open up a discussion of some ideas and also gather feedback and suggestions on what I think might be some possible solutions. Also, this discussion will list some essential items and highlight some pain points with Flutter version management.

## Principles

While maintaining the project, as it gained popularity, many individual requests have been to extend, enhance or modify some Flutter CLI behaviors. That is why I try to follow these principles.

- Always use Flutter tools when interacting with the SDK.
- Do not override any Flutter CLI commands.
- Follow Flutter suggested installation instructions accomplish caching.
- Should extend Flutter behavior and not modify it.
- API should be simple and easy to understand.

## Basic use cases

- Cache multiple Flutter versions (faster switching without download)
- Configure a version per project Reference
- Flutter version routing Reference
- IDE compatibility (vscode, android studio detects linked version)

## Challenges

As you can imagine, some challenges exist when supporting a tool like Flutter and the breadth of dev environments (devs love to customize their environments!)

### Routing

- The commands need to be proxied fvm flutter {command} to support routing. It passes the command and runs as a separate process. It works surprisingly well but sometimes can be frustrating from a DX perspective with an issue like #202
- Routing solves the monorepo problem and passes the Flutter SDK that is running the process down into its environment variables for any other sub-process to grab the same version.
- Routing is not "always" needed, and to be honest, I prefer to call the SDK directly when running standard build and run commands; when you have subprocesses, that is when routing comes in handy.

### Pub global activate

- One of the most significant pain points with FVM is that most users install with dart pub global activate; because most Dart SDK users are Flutter developers, it means the Dart SDK version is embedded with Flutter SDK, which means a change in the Flutter (PATH) global version would cause a difference in the Dart version, and cause the following issue [https://github.com/dart-lang/pub/issues/2593](https://github.com/dart-lang/pub/issues/2593)
- You can install FVM as an executable, Brew, Chocolatey, which solves the problem above.

## Suggestions

Based on my experience, here are some suggestions for changes that can be made to remove some of the noise added to the Flutter team and simplify FVM in some ways.

- Deprecate FVM global command. I understand the benefit of this command, but maintaining the main version of Flutter through the FVM causes confusion and noise when troubleshooting.
- The best practice is doing a standard Flutter install as your default/global SDK install and using FVM on a per-project basis. It allows for better compatibility and cleaner signals of what is happening behind the scenes.
- Make executable the default install recommended method for installing FVM.
- Implement version pinning in a more standard way. The Flutter team provides an "environment.flutter" pubspec parameter for pinning the version. It seems only to be supported by their internal tools; however, it would be an excellent way to provide a version pinning method that works across the community.
- Removing the `global` workflow, would open up for trimming down and simplification of FVM in general, and remove a lot of edge cases.
- Open for the community to maintain FVM-dependent tools, like GitHub Actions, Docker Images, etc.

If you get to this point, any feedback/suggestions would be appreciated it.

**Comment by EArminjon** [comment]
Little feedback from our team : we never used the global command from fvm.  We installed fvm from brew (MacOs) and use it allow the .fvm config file. It's a pleasure to count on this cli. Never got issue, well done :) !

(No issue on our runners too <3)

**Comment by rodrigorahman** [comment]
Hello, I particularly think the deprecated of the global command is very bad.

There is a huge advantage in being able to switch Flutter versions based on a specific version think like this:

I am working from home office for 2 companies and both are a consultancy where they develop systems for many companies and each of these companies has a project for each different version.

Imagine the difficulty or work I have to configure the version I need in ALL projects and if by chance I open the terminal and run a flutter pub get or I want to leave the terminal running build runner. I will always have to replace the main command with some fvm and remember that in the heat of development it always generates many problems or even strange bugs that make us lose hours.

I see global activation as a pattern that is used by other languages such as nodejs and this is a pattern that is widely used in the market. And the FVM does it very well.

In my opinion what should be changed is the documentation with only 1 way to install (Brew, choco), because in my experience problems always occur when we install the FVM as a package. Perhaps this would avoid many problems like the one mentioned.

And also have an explanatory video on how to do the installation step by step, thus leaving a standard for use and also a FAQ with troubleshooting, for example, whenever you change the flutter version you must clear the packages cache, and other problems acquaintances.

In my opinion, configuration per project is a complicated item that raises a lot of doubt for those starting out, as it is much easier to teach a DEV to run a fvm global stable, fvm global 2.7.0 command than to teach a step-by-step how to it configures the flutter version within a project and how it checks if it is in the same version. Experience from over 900 students and many of them beginners.

;-)

**Reply by EArminjon** [reply]
In our ci-cd we created the alias 'flutter' to 'fvm flutter'. We consider that using flutter without fvm is an error. Each of our projects contain the fvm config file.

Using flutter outside a project folder is an error for us.

**Reply by leoafarias** [reply]
@rodrigorahman, many challenges come with having FVM as a global installation. One way to think about it is that FVM is a pub package that depends on the Dart install it manages.

This causes many side effects that we have documented. Still, it is challenging because some are bugs/issues in the Flutter tools that need to be addressed, and most likely won't be because switching Flutter versions is not something they optimize for.

Some other challenges include "noise" issues, which are difficult to troubleshoot to see if they're caused by environment PATH configuration, and all FVM issues on the official Flutter repo are caused by the global setup. It's not that FVM has an issue, but it gets flagged because they're running through FVM. I have talked to the Flutter team, and this is one way to minimize this.

I am considering leaving the global command as a manual configuration and stating that it is an experimental and advanced feature.

The way I use FVM, which works well, is to have the latest stable globally and to have FVM configured per project. This simplifies the workflow and helps me troubleshoot any Flutter CLI issues.

Thank you for your feedback. Your input is valuable, but I want to emphasize some of our challenges.

**Reply by EArminjon** [reply]
> > I am considering leaving the global command as a manual configuration and stating that it is an experimental and advanced feature.
> 
> Please do keep it around. The `global` command is a huge timesaver for me. I never personally had issues with using it in the past 1.5 years since I started using fvm. Having to install and upgrade Flutter versions manually seems like a chore.

Question : What is your usage of fvm global ? I work since 3 years with flutter and manage more than 30 apps. I only use fvm and flutter inside project repository.  I never use the global.

**Reply by Auxx** [reply]
@EArminjon global Flutter is required when you're using global Flutter tools like [derry](https://pub.dev/packages/derry). It is impossible to use them through fvm.

Also prefixing every command with `fvm` is just plain wrong and creates a lot of confusion as no official docs mention it. All the docs for flutter and dart use commands directly. `fvm` should only be used to manage versions, not to proxy flutter or dart.

**Reply by leoafarias** [reply]
> @EArminjon global Flutter is required when you're using global Flutter tools like [derry](https://pub.dev/packages/derry). It is impossible to use them through fvm.
> 
> Also prefixing every command with `fvm` is just plain wrong and creates a lot of confusion as no official docs mention it. All the docs for flutter and dart use commands directly. `fvm` should only be used to manage versions, not to proxy flutter or dart.

Hey @Auxx, could you please elaborate on your use case? I'm interested in understanding the reason behind your question. FVM offers an "exec" command that lets you wrap the Flutter context, which I believe can solve the issue you're facing. If you provide me with more specific details about your use case, I can ensure that the solution I suggest will address your needs.

**Reply by Auxx** [reply]
@leoafarias yes, of course!

First of all, let me explain why we have a specific workflow. Our company mostly uses Java and Node.js and there are two version management systems for them: SDKMAN and NVM. They both work the same way by updating local environment to point to a specific Java or Node version.

Our approach relies on the following:

1. We use native tools for each language for development processes (Gradle for Java and NPM for Node) as much as possible.
2. Our codebase is the single source of truth for the whole development lifecycle. That means that all development, debugging, CI/CI, etc tasks and configs (except for sensitive data like keys) are part of the source code.

So imagine the following, we might have a set of `test` tasks with different arguments passed to test runners. In Java we will create gradle tasks like `testCi`, `testDev`, etc and we are able to run them using gradle directly: `$ gradle testCi`. The same approach is taken for Node projects and tasks are defined as NPM tasks, so then we can run them like: `$ npm run testCi`. We are also able to use the exact same commands on any environment, we don't have to prefix them and they look and feel exactly the same as they are mentioned in official documentation for these tools (Gradle and NPM).

We are now also started developing Flutter apps and we have adopted Derry to run tasks defined in `pubspec.yaml`. For example, we have a watch task, which is defined as follows: `watch: flutter packages pub run build_runner watch`. And we run it by `$ derry watch`.

So here are multiple issues with your approach to FVM:

1. How do we even install Derry correctly? It should be installed globally, but when switching between Dart/Flutter versions, cached Derry binary causes crashes on start up.
2. Derry tasks might execute Flutter or Dart CLI commands, these will never be invoked through with FVM.
3. CI/CD environments are usually set up against specific Flutter version and they don't use FVM, that means now we have to remember when to prefix our commands with fvm and when not to.
4. This all leads to a situation that we cannot have our source code as a single source of truth and we have to remember when to do X and when to do Y.
5. This also leads to a situation where all other languages used in our company work one way yet Flutter/Dart works in a different way.
6. And as a bonus fvm prefix is not present in any docs for any Flutter tools. That just adds additional mental complexity.

In my opinion, this means that FVM is not fit for purpose for us. It doesn't work the same and expected way as similar tools for other languages do, it adds complexity and doesn't really solve the main problem - versioning of Flutter SDK.

**Reply by kuhnroyal** [reply]
Just prefix the scripts with `fvm` and alias `fvm flutter` to `flutter` on CI, or the other way around?

**Reply by Auxx** [reply]
@kuhnroyal shell aliases are not part of the source code. That breaks our core principle and defeats the purpose of using task runners and other tools.

**Reply by kuhnroyal** [reply]
Shell aliases in any modern CI/CD configuration are of course part of the source code.

**Reply by Auxx** [reply]
@kuhnroyal they are not and they are not compatible with existing workflows. If FVM team is unwilling to switch to an industry compatible workflows, then we will search for alternatives.

**Reply by Isuru-Nanayakkara** [reply]
> > > I am considering leaving the global command as a manual configuration and stating that it is an experimental and advanced feature.
> > 
> > 
> > Please do keep it around. The `global` command is a huge timesaver for me. I never personally had issues with using it in the past 1.5 years since I started using fvm. Having to install and upgrade Flutter versions manually seems like a chore.
> 
> Question : What is your usage of fvm global ? I work since 3 years with flutter and manage more than 30 apps. I only use fvm and flutter inside project repository. I never use the global.

Without a global Flutter version, it's not possible to create a Flutter project by running `fvm flutter create`.

**Reply by EArminjon** [reply]
> > > > I am considering leaving the global command as a manual configuration and stating that it is an experimental and advanced feature.
> > > 
> > > 
> > > Please do keep it around. The `global` command is a huge timesaver for me. I never personally had issues with using it in the past 1.5 years since I started using fvm. Having to install and upgrade Flutter versions manually seems like a chore.
> > 
> > Question : What is your usage of fvm global ? I work since 3 years with flutter and manage more than 30 apps. I only use fvm and flutter inside project repository. I never use the global.
> 
> Without a global Flutter version, it's not possible to create a Flutter project by running `fvm flutter create`.

Create folder, fvm use 3.3.10 --force. fvm flutter create .

**Reply by Isuru-Nanayakkara** [reply]
> > > > > I am considering leaving the global command as a manual configuration and stating that it is an experimental and advanced feature.
> > > > 
> > > > 
> > > > Please do keep it around. The `global` command is a huge timesaver for me. I never personally had issues with using it in the past 1.5 years since I started using fvm. Having to install and upgrade Flutter versions manually seems like a chore.
> > > 
> > > 
> > > Question : What is your usage of fvm global ? I work since 3 years with flutter and manage more than 30 apps. I only use fvm and flutter inside project repository. I never use the global.
> > 
> > 
> > Without a global Flutter version, it's not possible to create a Flutter project by running `fvm flutter create`.
> 
> Create folder, fvm use 3.3.10 --force. fvm flutter create .

Then the `.fvm` directory gets created outside of the project directory, right. Do you keep it that way or move it inside manually?

**Reply by EArminjon** [reply]
if you did 
```bash
mkdir myapp
cd myapp
fvm use 3.3.10 --force
fvm flutter create .
```

fvm and flutter are in the same folder. No sub project folder.

**Reply by Isuru-Nanayakkara** [reply]
> if you did
> 
> ```shell
> mkdir myapp
> cd myapp
> fvm use 3.3.10 --force
> fvm flutter create .
> ```
> 
> fvm and flutter are in the same folder. No sub project folder.

Gotcha! I missed the period. This is helpful. Thank you!

**Reply by leoafarias** [reply]
> @leoafarias yes, of course!
> 
> First of all, let me explain why we have a specific workflow. Our company mostly uses Java and Node.js and there are two version management systems for them: SDKMAN and NVM. They both work the same way by updating local environment to point to a specific Java or Node version.
> 
> Our approach relies on the following:
> 
> 1. We use native tools for each language for development processes (Gradle for Java and NPM for Node) as much as possible.
> 2. Our codebase is the single source of truth for the whole development lifecycle. That means that all development, debugging, CI/CI, etc tasks and configs (except for sensitive data like keys) are part of the source code.
> 
> So imagine the following, we might have a set of `test` tasks with different arguments passed to test runners. In Java we will create gradle tasks like `testCi`, `testDev`, etc and we are able to run them using gradle directly: `$ gradle testCi`. The same approach is taken for Node projects and tasks are defined as NPM tasks, so then we can run them like: `$ npm run testCi`. We are also able to use the exact same commands on any environment, we don't have to prefix them and they look and feel exactly the same as they are mentioned in official documentation for these tools (Gradle and NPM).
> 
> We are now also started developing Flutter apps and we have adopted Derry to run tasks defined in `pubspec.yaml`. For example, we have a watch task, which is defined as follows: `watch: flutter packages pub run build_runner watch`. And we run it by `$ derry watch`.
> 
> So here are multiple issues with your approach to FVM:
> 
> 1. How do we even install Derry correctly? It should be installed globally, but when switching between Dart/Flutter versions, cached Derry binary causes crashes on start up.
> 2. Derry tasks might execute Flutter or Dart CLI commands, these will never be invoked through with FVM.
> 3. CI/CD environments are usually set up against specific Flutter version and they don't use FVM, that means now we have to remember when to prefix our commands with fvm and when not to.
> 4. This all leads to a situation that we cannot have our source code as a single source of truth and we have to remember when to do X and when to do Y.
> 5. This also leads to a situation where all other languages used in our company work one way yet Flutter/Dart works in a different way.
> 6. And as a bonus fvm prefix is not present in any docs for any Flutter tools. That just adds additional mental complexity.
> 
> In my opinion, this means that FVM is not fit for purpose for us. It doesn't work the same and expected way as similar tools for other languages do, it adds complexity and doesn't really solve the main problem - versioning of Flutter SDK.

Thanks for explaining your thoughts, @Auxx! It's important to note that FVM isn't trying to be a perfect tool for everyone using Flutter or meet every possible workflow. The main challenge with FVM is that Flutter doesn't make it easy to have multiple versions on the same machine, especially when it comes to pub dependencies. This means that FVM won't work like other tools such as NVM. However, I think the experience is close to it, and that's why I recommend moving global, which would allow for a clear version that manages the pub.dev dependencies.

While there may be workarounds for the issues you mentioned, such as using "alias" or the "exec" command, they may need to be optimized for your workflow. If you have specific use cases with Derry, feel free to open up an issue so we can take a look. Most of the other issues are out of FVM's control, to be honest.

**Comment by adamhalesworth** [comment]
I agree with @rodrigorahman. One of the benefits of fvm is that it works in a similar way to other tools such as nvm. Removing the option to set a global version might equally cause confusion (which is what led me here).

If this decision stems mostly/only from the perspective of troubleshooting a "non-standard" install, removing the global workflow to remedy this feels excessive and reduces fvm's usefulness.

Richer documentation or improvements to cli DX would be far more beneficial than deprecating core functionality.

**Reply by leoafarias** [reply]
@adamhalesworth, as I mentioned in my reply to @rodrigorahman, Flutter CLI tools are not optimized for version switching. I have done my best to provide context and messages when running commands that might cause issues, but the biggest challenge is that FVM depends on the global install, creating a circular dependency that can cause issues. All of these issues have fixes, but there is too much noise that comes from it.

**Reply by adamhalesworth** [reply]
Thank you for further highlighting the challenges on your side @leoafarias. There are so many moving parts to take into consideration, I know maintenance on fvm will not be easy and that these are difficult decisions to make.

Recommending the executable install method and retaining `global` as an advanced feature sounds like the ideal compromise. I have always installed fvm via brew which is likely why I have not come across the issues myself.

**Reply by leoafarias** [reply]
Since this has come up in another conversation. I believe there is plan forward for a  simpler "global" version. That addresses the concerns while also simplifying the workflow.

**Comment by leoafarias** [comment]
Hello! FVM 3.0 is nearing completion and we're opening up Beta testing. If you're interested in hands-on testing and direct discussions about issues or use-cases, please sign up.
@adamhalesworth @Isuru-Nanayakkara @EArminjon @rodrigorahman @Auxx @kuhnroyal 

https://forms.gle/n2Qm6zHX9aLc14gs7

**Reply by ManuelRauber** [reply]
Hi @leoafarias 

I've seen that you will deprecate the global version with FVM 3.0.

That's unfortunate, because that's a thing people are using. 

While I normally only work with local installs if possible, I've a project that require a global flutter installation.

The project uses a 3rd party that comes with a shell script. That shell script will directly call `flutter ...` and thus will use the global version. 

With whatever I read in the FVM 3.0 sources the official way will installing the Flutter SDK globally and as well via FVM.

However, in that case there is no reason to use FVM 3.0 for that particular project, because I need to install the Flutter SDK globally anyway.

Now., the twist is: I have two projects with that 3rd party thing and they require different Flutter frameworks. 

At the moment, I run `fvm global` before each project and it works as expected. 

Do you have any other idea how to solve that?

**Reply by leoafarias** [reply]
Global will not be deprecated anymore. However I will provide guidance as I believe the overuse of global means most likely there is an incorrect use by a lot of people.

Version per project versioning should be thought as per project "global" we should have a beta this week.
